### PR TITLE
Simplify onboarding code, remove possible bugs

### PR DESCRIPTION
**Pull request by asutula**
The approach here was to clean up and simplify:

- Remove band aid fixes that haven't worked
- Remove array splicing based on redux state, always show referral code screen, pre-populated with a referral code when applicable
- Remove a couple things that might have been causing bugs

### CODE REVIEW

File: App/Components/ReferralCode.tsx [hunk 1]
```diff
@@ -59,7 +59,8 @@ const HIT_SLOP: Insets = {
 }
 
 interface Props {
-  referralCode: string
+  targetReferralCode: string
+  referralCode?: string
```
Comment (asutula) [inline App/Components/ReferralCode.tsx]: The referral code input component now optionally takes a prop that is a referral code (set in the case the user is onboarding from opening the app from an invite link)

File: App/Containers/OnboardingScreen/index.tsx [hunk 2]
```diff
@@ -29,7 +29,7 @@ const ARROW_FORWARD: ViewStyle = {
 }
 
 interface StateProps {
-  skipReferralCode: boolean
+  referralCode?: string
```
Comment (asutula) [inline App/Containers/OnboardingScreen/index.tsx]: No more skipping referral code component

File: App/Containers/OnboardingScreen/index.tsx [hunk 3]
```diff
@@ -39,7 +39,6 @@ interface DispatchProps {
 type Props = StateProps & DispatchProps & NavigationScreenProps
 
 interface State {
-  blockNext: boolean
```
Comment (asutula) [inline App/Containers/OnboardingScreen/index.tsx]: Removing this fix that didn't seem to help

File: App/Containers/OnboardingScreen/index.tsx [hunk 4]
```diff
@@ -72,19 +70,20 @@ class OnboardingScreen extends React.Component<Props, State> {
...
+        showArrow: this.showArrowForIndex(this.state.currentPage + 1)
+      })
       this.pages.scrollToPage(this.state.currentPage + 1)
-      setTimeout(() => {
-        this.setState({
-          blockNext: false,
-          showArrow: this.showArrowForIndex(this.state.currentPage + 1),
-          currentPage: this.state.currentPage + 1
-        })
-      }, 350)
```
Comment (asutula) [inline App/Containers/OnboardingScreen/index.tsx]: `nextPage` now simply tells the pager component to scroll to the next page and sets the arrow visibility based on what the next page will be.

File: App/Containers/OnboardingScreen/index.tsx [hunk 5]
```diff
@@ -72,19 +70,20 @@ class OnboardingScreen extends React.Component<Props, State> {
...
     }
   }
 
+  onScrollEnd = () => {
+    this.setState({
+      currentPage: this.state.currentPage + 1
+    })
+  }
+
```
Comment (asutula) [inline App/Containers/OnboardingScreen/index.tsx]: Using the `onScrollEnd` callback from the pager component to update `state.currentPage` once we've actually arrived on that page.

File: App/Containers/OnboardingScreen/index.tsx [hunk 6]
```diff
@@ -149,10 +149,6 @@ class OnboardingScreen extends React.Component<Props, State> {
         />
       )
     ]
-    // remove referral code if needed.
-    if (this.props.skipReferralCode) {
-      pages.splice(3, 1)
-    }
```
Comment (asutula) [inline App/Containers/OnboardingScreen/index.tsx]: No more dynamic splicing of the page data. It it always the same now.

Comment (carsonfarmer) [reply App/Containers/OnboardingScreen/index.tsx]: 👍

File: App/Containers/OnboardingScreen/index.tsx [hunk 7]
```diff
@@ -166,8 +162,8 @@ class OnboardingScreen extends React.Component<Props, State> {
             style={[CONTAINER]}
             containerStyle={{ marginBottom: spacing._016 }}
             indicatorColor={color.accent2_2}
-            startPage={this.state.currentPage}
```
Comment (asutula) [inline App/Containers/OnboardingScreen/index.tsx]: This is actually the thing I'm most suspicious could have caused problems. This `render` method gets called over and over and over again. `startPage` is, in theory, something that should be static, but we were setting it to `state.currentPage` which is changing as the user progresses through the onboarding. You would think that the pager component should deal with this correctly, but there's a good change it doesn't and may be causing our bug.

Comment (asutula) [reply App/Containers/OnboardingScreen/index.tsx]: So, omitting this prop, it defaults to `0`, and we never pass in any other value.

**Pull request merged**